daemons/ipa-kdb/ipa_kdb_principals.c

```diff
@@ -338,25 +337,75 @@ static void ipadb_validate_otp(struct ipadb_context *ipactx,
     if (dn == NULL)
         return;
     count = asprintf(&filter, ftmpl, dn, datetime, datetime);
-    ldap_memfree(dn);
-    if (count < 0)
+    if (count < 0) {
+        ldap_memfree(dn);
         return;
+    }
 
     /* Fetch the active token list. */
     kerr = ipadb_simple_search(ipactx, ipactx->base, LDAP_SCOPE_SUBTREE,
                                filter, (char**) attrs, &res);
     free(filter);
+    filter = NULL;
     if (kerr != 0 || res == NULL)
         return;
 
     /* Count the number of active tokens. */
     count = ldap_count_entries(ipactx->lcontext, res);
     ldap_msgfree(res);
 
-    /* If the user is configured for OTP, but has no active tokens, remove
-     * OTP from the list since the user obviously can't log in this way. */
-    if (count == 0)
+    /*
+     * If there are no valid tokens then we need to remove the OTP flag,
+     * unless OTP is the only auth type allowed...
+     */
+    if (count == 0) {
+        /* Remove the OTP flag for now */
         *ua &= ~IPADB_USER_AUTH_OTP;
+
+        if (*ua == 0) {
+            /*
+             * Ok, we "only" allow OTP, so if there is an expired token then
+             * add back the OTP flag as the server will double check the
+             * validity and reject the entire bind. Otherwise, this is the
+             * first time the user is authenticating and the user should be
+             * allowed to bind using its password
+             */
+            static const char *expired_ftmpl = "(&"
```

Comment:
We also can have a token disabled. It would mean that a token is not expired but still not available for use. I think disabled token should be treated as expired one too?

Reply:
I agree, the only time the authentication should fall through to password is if the user has NEVER used OTP to authenticate.  So expired/disabled token means it had a OTP token at some point.  <just merged changed>